### PR TITLE
realtek: Merge new realtek features

### PR DESCRIPTION
**Pull request by ghost**
Merge several new features and fix bugs in the ethernet driver.
New features are support for offloading TC Flower filters into the RTL SoC's Packet inspection engine and offloading of L3 routing.
Also new is support for the 10G AQR113C PHY as found on Zyxel switches with RTL9300 SoC.

### CODE REVIEW

**Comment by bkobl** [comment]
I just noticed I had forgotten to add the new file dsa/tc.c, which provides the TC flower implementation.

**Comment by Borromini** [comment]
@bkobl Just flashed master with this PR, seeing this in dmesg, not sure if this is an issue:

```
root@OpenWrt:~# dmesg -l err
[    8.996810] rtl83xx_fib_event_work_do: FIB4 default rule failed
[   25.671028] rtl83xx-switch switch@bb000000: unknown nexthop, id 0
[   25.720590] rtl83xx-switch switch@bb000000: unknown nexthop, id 0
[   25.734187] rtl83xx-switch switch@bb000000: unknown nexthop, id 0
[   25.741152] rtl83xx_fib4_del: no such gateway: 0.0.0.0
[   46.241245] rtl83xx_fib_event_work_do: FIB4 default rule failed
[   48.079263] rtl83xx_fib_event_work_do: FIB4 default rule failed
```

**Comment by bkobl** [comment]
This should not be an issue. The HW of the RTL83xx is not able to handle default routes whee the gateway is 0.0.0.0. So they are not added. Then the kernel tries to delete it again when going from the rescue network setup to the one configured in /etc/config/network, which fails. There are maybe a bit more warning messages in that code than necessary, but if we find bugs this should help to identify them.

**Comment by bkobl** [comment]
Routes the HW supports are e.g.
ip route add 192.168.4.1 via 192.168.100.150
everything else should be ignored for offloading, which will give warnings currently.

**Comment by bkobl** [comment]
If you want to test tc flower, you need several additional modules linked in, I use (probably much more than strictly necessary):
iptable_nat ipt_REJECT xt_time xt_tcpudp xt_string xt_state xt_recent xt_quota xt_pkttype xt_owner xt_nat xt_multiport xt_mark xt_mac xt_limit xt_helper xt_conntrack xt_connmark xt_connlimit xt_connlabel xt_connbytes xt_comment xt_cgroup xt_bpf xt_addrtype xt_TCPMSS xt_REDIRECT xt_MASQUERADE xt_LOG xt_CT ts_kmp ts_fsm ts_bm sch_mqprio sch_cake nf_reject_ipv4 nf_nat nf_log_ipv4 nf_conntrack_netlink nf_conncount iptable_raw iptable_mangle iptable_filter ip_tables cls_flower act_vlan sch_teql sch_sfq sch_red sch_prio sch_pie sch_multiq sch_gred sch_fq sch_dsmark sch_codel em_text em_nbyte em_meta em_cmp act_simple act_police act_pedit act_ipt act_csum libcrc32c em_ipset cls_bpf act_bpf act_ctinfo act_connmark nf_conntrack nf_defrag_ipv6 nf_defrag_ipv4 sch_tbf sch_ingress sch_htb sch_hfsc em_u32 cls_u32 cls_tcindex cls_route cls_matchall cls_fw cls_flow cls_basic act_skbedit act_mirred act_gact xt_set ip_set_list_set ip_set_hash_netportnet ip_set_hash_netport  ip_set_hash_netnet ip_set_hash_netiface ip_set_hash_net ip_set_hash_mac ip_set_hash_ipportnet ip_set_hash_ipportip ip_set_hash_ipport ip_set_hash_ipmark ip_set_hash_ip ip_set_bitmap_port ip_set_bitmap_ipmac ip_set_bitmap_ip ip_set nfnetlink nf_log_ipv6 nf_log_common ip6table_mangle ip6table_filter ip6_tables ip6t_REJECT x_tables nf_reject_ipv6 gpio_button_hotplug crc32c_generic

testing then is done via e.g.
 # Test: QDISC
tc qdisc add dev eth0 ingress

tc filter add dev eth0 protocol ip parent ffff: handle 2 flower src_ip 192.168.2.150 skip_sw action drop

tc filter add dev eth0 protocol ip parent ffff: handle 2 flower src_mac 24:5e:be:50:cb:98 skip_sw action drop

tc filter show dev eth0 ingress
tc filter del dev eth0 ingress
tc -s filter show dev eth0 ingress

# Redirection
tc filter add dev eth0 protocol ip parent ffff: handle 2 flower src_ip 192.168.2.150 skip_sw action mirred egress redirect dev lan3

tc filter add dev eth0 protocol ip parent ffff: handle 2 flower src_ip 192.168.2.150 skip_sw action mirred egress mirror dev lan3

# VLAN
tc filter add dev eth0 parent ffff: protocol 802.1Q flower vlan_id 100 skip_sw action drop

tc filter add dev eth0 parent ffff: protocol 802.1Q flower vlan_id 100 skip_sw action vlan pop

tc filter add dev eth0 parent ffff: protocol 802.1Q flower vlan_id 100 skip_sw action vlan push id 200

tc filter add dev eth0 protocol ip parent ffff: handle 2 flower dst_ip 192.168.2.150 skip_sw action vlan push id 100

**Comment by Neustradamus** [comment]
@bkobl: Good job, can not wait a merging ^^

**Comment by bkobl** [comment]
Added a bug fix that was discussed in the forum: https://forum.openwrt.org/t/support-for-rtl838x-based-managed-switches/57875/733

**Comment by bobafetthotmail** [comment]
This is very good. HW routing support is very important for these devices.

Afaik Aquantia AQR113C PHY aren't specific to realtek target, they can be used on x86 target as well at the very least.
I think that patch should be moved to the generic kernel patches so all targets can use it if needed.
Also since it's not required for booting it's probably better to turn it in a package, see https://github.com/openwrt/openwrt/commit/b63e57ba1795dfedd425d5cd6fffdae56274b3d1

**Comment by bkobl** [comment]
Indeed, I will add the Aquantia patch to the general kernel patches. But making it a packet I am not sure, it is merely an additional feature of an existing module, hopefully to go upstream at some point.

**Comment by adschm** [comment]
Since we have testing support for 5.10 on realtek now, you might consider to keep things easier by just applying these changes to 5.10. (However, note that _I_ won't merge this PR since it's out of my league (unless you have a pile of Reviewed-by ...))

**Comment by bkobl** [comment]
Indeed, this should now be applied to 5.10. The biggest issue is that the 5.10 kernel breaks the GPIO support for 93xx SoCs, it is simply not implemented there. I am hoping that @svanheule who wrote the driver in 5.10 can fix this. At the moment I am working with @BrainSlayer on the 931x support under 5.10 but we just use the 5.4 GPIO driver (RTL930x and RTL931x are identical for GPIO). If it is too much effort to port this PR to 5.10 then we'll add the features together with 931x support probably sometime in 2022. Maybe @blogic can give some advice.

**Comment by BrainSlayer** [comment]
i dont think that the otto gpio driver is hard to fix. i can handle this if required. but the gpio has no high priority right now, except for i2c usage (fan control etc.). i havent reviewed the code yet.

**Comment by BrainSlayer** [comment]
btw. you can find a 5.10 based tree for all common supported switches in my github account. there is a pie tree with several branches. the 5.10 branch is tested on 839x devices and includes all current works from me and bkobl. the 5.10-rtl9313 branch does  also support the older switch models but hasnt been tested. its a refactored enhanced variant of the 5.10 patch and has several different compiler profiles for performance reasons. the original openwrt code does compile 839x and newer switches using mips 4kec instructions. but the cpu is mips34kc in fact. so the original 5.10 and also 5.4 support of openwrt does not run with full performance and lacks several mips related features on some models liks mips16, mips-mt smp support etc. this has been fixed in that tree.

**Comment by svanheule** [comment]
I think I mentioned this before, but you should already be able to use the realtek-otto GPIO driver, with the fallback compatible.
If you used ```compatible = "realtek,rtl9300-gpio", "realtek,otto-gpio";```, the first compatible would be ignored until it's implemented. The RTL93xx SoC implement their GPIO banks in a different layout however, so you would need to enable all 32 GPIOs, even if they aren't available. GPIO pins 0..7 would then have index 24..31, 8..15 mapping to 16..23, etc.
The interrupts aren't supported on the fallback compatible, because that would cause trouble if you aren't using the exact correct port layout.

**Comment by BrainSlayer** [comment]
@svanheule yes looks like it might work if the dts uses the correct mem offset. irq support isnt a big deal. looks easy to fix.

**Comment by bkobl** [comment]
I am just rebasing to move this to 5.10 and it looks good so far. I wanted to change the patch adding support for the Aquantia AQR113C to be a generic 5.10 kernel patch as suggested by Alberto (and throw in support for AQR813 as found on the Zyxel XS1930-10 while at it). But where should this patch go? target/linux/generic/hack-5.10 or pending-5.10? Can it stay in this PR?

**Comment by adschm** [comment]
Just answering generally, pending-5.10 if it has been submitted somewhere, or hack-5.10 if it's just here.

**Comment by bkobl** [comment]
Could someone help me with this PR? I would like to push only the stuff from
realtek: Correct TX ring size in ethernet driver 
onwards. How do I do that?

**Comment by bkobl** [comment]
OK, so I managed this. The PR is now for 5.10.

**Comment by lhuedepohl** [comment]
Just FYI: I tested c6f6113 on both a DGS-1210-28 and a DGS-1210-26, and with the included bug-fixes the switch is now finally able to forward packets between ports, thanks a lot!

**Comment by blogic** [comment]
tested on 4 different devices. IMHO this is good to go, if there are no big doubts, i'd merde this.
@bkobl @svanheule @bmork @adschm

**Comment by adschm** [comment]
No objections, but I'm not much help here.

File: target/linux/realtek/dts-5.10/rtl838x.dtsi [hunk 1]
```diff
@@ -148,7 +148,7 @@
 		};
 
 		gpio0: gpio-controller@3500 {
-			compatible = "realtek,rtl8380-gpio", "realtek,otto-gpio";
+			compatible = "realtek,otto-gpio";
```
Comment (svanheule) [inline target/linux/realtek/dts-5.10/rtl838x.dtsi:151]: This change is a downgrade, you need both compatibles.

If you only leave the "realtek,otto-gpio" compatible, you lose GPIO interrupts, which is useful for efficiently registering button presses.

**Review by svanheule — COMMENTED** [review]
Regarding the reintroduction of the original GPIO driver: I've indicated how I think you could switch to the fallback implementation. What is not working with the realtek,otto-gpio driver that requires you to reintroduce the old driver?

**Comment by svanheule** [comment]
I've left a few comments on the GPIO related stuff, but that's about all I can give useful feedback on. I think it may be better if these big pull request were split up into smaller PRs with better defined topics, in the hopes of attracting attention from people who know something about that specific topic.

**Comment by bkobl** [comment]
The new realtek,otto-gpio GPIO driver does not work on RTL9300 SoCs, therefore the re-introduction of the old one. That the new one re-uses the old compatibility string with a new meaning is a real mess. For a non-testing kernel, the cleaner solution would be to use only the old driver until the new one is fixed.

With regards to splitting this PR: This PR adds support for the Routing offloading, which is really a feature of the RTL9300 SoCs. Also people have been crying to finally get cheap Multi-Gig routers on OpenWRT. And in order to test this, all RTL9300 features need to be there. There is no way to test real devices without e.g. the GPIO driver working. It would not make sense to submit a PR that people cannot test. Also, there were hardly ever any comments on the ethernet and switch drivers of the SoCs, so I don't think separating this out would help.

Also, this is a PR against the testing kernel, if there are things to be ironed out, like making the new GPIO driver work for the RTL9300, there is still time. Note that I supported that the new GPIO driver would be put into the testing kernel even though it was breaking RTL9300 support. What really needs testing here is all the offloading/Multi-Gig code. People want to be able to use these features because without them the routing and firewalling performance is very limited for these devices with their very limited CPU. Getting these features in has already been considerably delayed, the next huge bunch is already coming up with the RTL9310-devices. Things like OpenFlow, SMP, GIC interrupts are partially working already and I would expect this to be ready for OpenWRT before Christmas now. But all of this sits on top of this PR as Sebastian and I started with it as a base due to the similarities of RTL9310 to the RTL9300.

**Comment by bkobl** [comment]
I dropped the commit that was downgrading the GPIO compatibility string, maybe this is a way forward.

**Comment by svanheule** [comment]
I was just trying to address the lack of review comments, but I guess this is a more general issue with OpenWrt due to limited resources. And then splitting things up in smaller pieces may not help anyhow. You're right that the sooner we can get things tested, the better. If @blogic hasn't  run into any issues, this is fine for me.

The GPIO stuff is just the part I know something about, because I worked on it. @musashino205 has submitted patches to use the GPIO interrupts, so that's being used now. I think you have a different DT-compatible for the original driver now, so that resolves the conflict you mentioned?

**Comment by bkobl** [comment]
Yes, I thought the compatibles were the same, but there is a 8380 in the new one and the old one uses 838x, so it is fine.

**Comment by musashino205** [comment]
BTW, old GPIO driver (current driver in 5.4) doesn't seem to support RTL93xx...

https://github.com/openwrt/openwrt/blob/d69bf6601e5bff897781361f2718513d4dd8efd8/target/linux/realtek/files-5.4/drivers/gpio/gpio-rtl838x.c#L335-L357

**Comment by bkobl** [comment]
Yes, the one I added to the PR takes the 9300 into account, however. It takes a long time for developed stuff to get into OpenWRT (see above), the 9300 GPIO driver did not make it at the time the rest of the 9300 code went in, which was in spring IIRC.

**Comment by musashino205** [comment]
Ahhh, Isee, I overlooked the changes for GPIO driver in this PR, sorry...

**Comment by blogic** [comment]
merged into master, Thanks !

**Comment by Neustradamus** [comment]
Thanks @bkobl and @blogic!

Merged commits are:
- https://github.com/openwrt/openwrt/commit/e88ac0bbd1099cc7844ad0efe891b9fb5f0e34e2
- https://github.com/openwrt/openwrt/commit/ca04e2dfd02a711cb0b3ce401bd120bd3da3b973
- https://github.com/openwrt/openwrt/commit/9eab76c84e311605e8cc197a0bc4bd53340678c3
- https://github.com/openwrt/openwrt/commit/2ab4d4029386849664d152949188c40590cfdd5f
- https://github.com/openwrt/openwrt/commit/a6678accbd3da00a002158618125af155cc27b37
- https://github.com/openwrt/openwrt/commit/54805fc91186228eec0ce14f7c5a6b4149b0a244
- https://github.com/openwrt/openwrt/commit/87b3bd07a445376d9cae73aaf9354029260183ac
- https://github.com/openwrt/openwrt/commit/2d8d81fe2897586b5ca0e78997c448899ae1e8a5
- https://github.com/openwrt/openwrt/commit/a96b73a8902854b4e810c4c24e1f0c35326722a4
- https://github.com/openwrt/openwrt/commit/9d9bf16aa8d966834ac1280f96c37d22552c33d1
- https://github.com/openwrt/openwrt/commit/28e972b2ea2f55a593defcbd2dc21710cce648c7
- https://github.com/openwrt/openwrt/commit/9ae927febd4b07e064bdb9eaeb43862ea6c7ac8f
- https://github.com/openwrt/openwrt/commit/ee6f483a625d4e7c6198d422dab1cfddf1de040a
- https://github.com/openwrt/openwrt/commit/35b3c4c212588deb49369635571ef63a7413b6f6
- https://github.com/openwrt/openwrt/commit/03e1d93e07791833e3167d8cfaef2e40a7113a23
- https://github.com/openwrt/openwrt/commit/1f402512ae0474ffd03c34bcfa54b2882d927861
- https://github.com/openwrt/openwrt/commit/1cfd45ae0bad3975cd59ff18ce04265b1bccbcd6
- https://github.com/openwrt/openwrt/commit/88936e7e8278ef82c0937e47be6b4556b3ef708a
- https://github.com/openwrt/openwrt/commit/76f60470a192f8ce585ac3289d3037b54d98ba11
- https://github.com/openwrt/openwrt/commit/4c83dae801a83fdfc07fce8aa68978839a76ae02
- https://github.com/openwrt/openwrt/commit/cda0460ad333f15e15ec9bd2516b2ae45f8a4644

**Pull request closed without merging**